### ISSUE TITLE
Document cmake/make/make install recipe for in-house monolithic code base

### ISSUE THREAD

**Issue by mristin** [conan-io/docs#716]
Hi conan team!

While we were trying to migrate the compilation and releasing of our (still monolithic) code base to conan, we were a bit confused how to use it with cmake wrapper from within a python build script.

Our usual work flow included:

* cd to the build directory, cmake {src} 
* make
* make check
* make install

We needed the "install" step so that the rpaths are removed, while we required the rpaths to be specified in the build phase to facilitate the unit testing (otherwise `make check` fails).

We looked at the documentation at https://docs.conan.io/en/latest/howtos/cmake_install.html. Unfortunately, it did not really help us further as this work flow was nowhere documented in full.

Here is what we come up with for the `conanfile.py`:

```python
import conans
import conans.tools


class Pqry(conans.ConanFile):
    name = "Pqry"
    version = "1.0.0"
    settings = "os", "compiler", "build_type", "arch"
    options = {"shared": [True, False]}
    generators = "cmake"
    keep_imports = True

    requires = (
        # some libraries here...
    )

    default_options = ("shared=True",
                       # some parameters here...
                       )

    def build(self):
        cmake = conans.CMake(self)
        cmake.configure()
        cmake.build()

    def package(self):
        cmake = conans.CMake(self)
        cmake.install()
```

In our python build script (which is our main point of entry), we wrote:

```python
import conans

def install_codebase(
    repo_dir: pathlib.Path, 
    build_dir: pathlib.Path, 
    release_dir: pathlib.Path) -> None:
    """
    Installs the Parquery codebase.

    :param repo_dir: where Parquery repository resides
    :param build_dir: where the build files will be stored
    :return:
    """
    source_dir = repo_dir / "src" / "cpp"
    cmake_build_debug_dir = source_dir / "cmake-build-debug"
    cmake_build_debug_dir.mkdir(exist_ok=True)
    build_dir.mkdir(exist_ok=True, parents=True)

    conan, _, _ = conans.client.conan_api.ConanAPIV1.factory()

    # Install files for clion
    conan.install(cwd=cmake_build_debug_dir.as_posix(), path="..", build=["missing"])

    # Install dependencies and build in the build directory
    conan.install(cwd=build_dir.as_posix(), path=source_dir.as_posix(), build=["missing"])

    # Specifying the argument 'package_folder' is necessary 
    # since it needs to be set at the build time for CMake.
    conan.build(cwd=build_dir.as_posix(),
                conanfile_path=source_dir.as_posix(),
                package_folder=release_dir.as_posix())

    conan.package(build_folder=build_dir.as_posix(),
                  path=source_dir.as_posix(),
                  package_folder=release_dir.as_posix())
```

Unfortunately, we could not make the `make check`. We finally resorted to:

```python
def run_make_check(build_dir: pathlib.Path) -> bool:
    """
    Runs 'make check' (unit tests for c++)

    :param build_dir: directory of the build-cpp files
    :return:
    """
    binaries_dir = build_dir / "bin"
    proc = subprocess.Popen(
        "PATH=$PATH:{} make check".format(binaries_dir),
        stderr=subprocess.PIPE,
        stdout=subprocess.PIPE,
        cwd=build_dir.as_posix(),
        shell=True)

    out, error = proc.communicate()

    if proc.returncode != 0:
        say("cmake check failed:\n{}\n{}".format(out, error))
        return False

    return True
```
, but this seems like a hack to us. Is there a way how we could invoke the `make check` with the conan client?

We would like to fix the documentation, but I wanted to see with you how you would like to document the following points that confused us:

* Re-creating the `cmake` object in `package()` is more intuitive to us than leaving it empty, as you suggest in https://docs.conan.io/en/latest/howtos/cmake_install.html. 

* The release directory (_i.e._ CMake install prefix) needs to be specified both at `build()` and `package()` calls.

* It would be great if we could document this somewhere as a general recipe for other people.

* We could not find anywhere the documentation for https://github.com/conan-io/conan/blob/develop/conans/client/conan_api.py. Is it available somewhere? It would be great to include the links to python methods from within your current documentation.

* Any help or documentation how to run `make check` is highly welcome!

Thanks in advance for sharing your thoughts and please let us know if and how you would like to split this issue into multiple ones!

**Comment by memsharded** [comment]
Hi @gizmomogwai https://docs.conan.io/en/latest/reference/build_helpers/cmake.html#build!

Thanks for your detailed feedback!

I am not sure about what you mean, maybe something in the line of?:

```python
def build(self):
        cmake = conans.CMake(self)
        cmake.configure()
        cmake.build()
        self.run("make check")
```
conan recipes are python scripts, you can run whatever you want there. With the ``self.run()`` method you are basically launching a subprocess, did you check: https://docs.conan.io/en/latest/reference/conanfile/other.html#running-commands?

The ``cmake.build()`` allows running specific targets, I haven't tried, but maybe this could work:

```python
def build(self):
        cmake = conans.CMake(self)
        cmake.configure()
        cmake.build()
        cmake.build(target="check")
```

> Re-creating the cmake object in package() is more intuitive to us than leaving it empty, as you suggest in https://docs.conan.io/en/latest/howtos/cmake_install.html.

I agree and it seems the docs in https://docs.conan.io/en/latest/howtos/cmake_install.html are not updated. If you want to contribute it in this page that would be great.

> The release directory (i.e. CMake install prefix) needs to be specified both at build() and package() calls.

Not sure what you mean, the ``CMAKE_INSTALL_PREFIX`` is already defined by the ``CMake`` helper automatically, you don't need to define it.

> It would be great if we could document this somewhere as a general recipe for other people.

Probably the best would be in https://docs.conan.io/en/latest/reference/build_helpers/cmake.html. If necessary to add a subsection called "Examples", that would be ok.

> We could not find anywhere the documentation for https://github.com/conan-io/conan/blob/develop/conans/client/conan_api.py. Is it available somewhere? It would be great to include the links to python methods from within your current documentation.

There is no documentation existing for the python API. We didn't want to document it yet, at least in our main docs, as that would be another thing to support, and it has been changing a lot, it is not stable yet, and the overhead would be considerable. We are considering adding to somewhere some code docs, that would be automatically generated.

Please let me know if the above helps. Thanks!

**Comment by mristin** [comment]
Hi @memsharded 
Thanks for your response! We will document the examples in the coming weeks.

> >  The release directory (i.e. CMake install prefix) needs to be specified both at build() and package() calls.

> Not sure what you mean, the CMAKE_INSTALL_PREFIX is already defined by the CMake helper automatically, you don't need to define it.

I was referring to:

```python3
    # Specifying the argument 'package_folder' is necessary 
    # since it needs to be set at the build time for CMake.
    conan.build(cwd=build_dir.as_posix(),
                conanfile_path=source_dir.as_posix(),
                package_folder=release_dir.as_posix())

    conan.package(build_folder=build_dir.as_posix(),
                  path=source_dir.as_posix(),
                  package_folder=release_dir.as_posix())
```
`package_folder` needs to be specfied both in the `conan.build()` call and in the `conan.package()` call. I would have expected to specify it only in the `conan.package()` call.

**Comment by memsharded** [comment]
Ah, now I get it.

The thing is that the cmake configuration step is called in the ``build()`` method (``cmake.configure()``), and that configuration already needs and caches CMAKE_INSTALL_PREFIX. The later invocation of cmake in the ``package()`` (``cmake.build()``, and ``cmake.install()``) method reuses the cached CMAKE_INSTALL_PREFIX value. It is more consistent to pass it, just in case, but it might happen that it could even work without defining it in your ``conan.package()`` call. It is that way because of the cmake design, I don't think this can be other way.

**Comment by danimtb** [comment]
This was fixed in #750
